src/runtime-rs/crates/shim/src/shim.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    os::unix::ffi::OsStrExt,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{anyhow, Context, Result};
+use sha2::Digest;
+
+use crate::{Args, Error};
+
+const SOCKET_ROOT: &str = "/run/containerd";
+const SHIM_PID_FILE: &str = "shim.pid";
+
+pub(crate) const ENV_KATA_RUNTIME_BIND_FD: &str = "KATA_RUNTIME_BIND_FD";
+
+/// Command executor for shim.
+pub struct ShimExecutor {
+    pub(crate) args: Args,
+}
+
+impl ShimExecutor {
+    /// Create a new instance of [`Shim`].
+    pub fn new(args: Args) -> Self {
+        ShimExecutor { args }
+    }
+
+    pub(crate) fn load_oci_spec(&self) -> Result<oci::Spec> {
+        let bundle_path = self.get_bundle_path()?;
+        let spec_file = bundle_path.join("config.json");
+
+        oci::Spec::load(spec_file.to_str().unwrap_or_default()).context("load spec")
+    }
+
+    pub(crate) fn write_address(&self, address: &Path) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("address");
+        std::fs::write(file_path, address.as_os_str().as_bytes())
+            .context(Error::FileWrite(format!("{:?}", &file_path)))
+    }
+
+    pub(crate) fn write_pid_file(&self, pid: u32) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join(SHIM_PID_FILE);
+        std::fs::write(file_path, format!("{}", pid))
+            .context(Error::FileWrite(format!("{:?}", &file_path)))
+    }
+
+    pub(crate) fn read_pid_file(&self, bundle_path: &Path) -> Result<u32> {
+        let file_path = bundle_path.join(SHIM_PID_FILE);
```

Comment:
Can you add a comment explaining why this method has a `bundle_path` param but all the others call `self.get_bundle_path()`.

Reply:
Nice catch @jodh-intel
By comparing with `write_pid_file`, the `bundle_path` can be retrieved by `self.get_bundle_path()?` instead of passed in.

Since this PR has taken too long, I'd like to merge this first and fix this in a new PR.

I have created an [issue](https://github.com/kata-containers/kata-containers/issues/3876) to track this.

Reply:
Thanks for @jodh-intel @liubin, since the PR has been merged, I will fix this in a new PR